AzureCore/Source/Pipeline/Policies/LoggingPolicy.swift

```diff
@@ -0,0 +1,133 @@
+//
+
+import Foundation
+
+public class DefaultLoggingPolicy: PipelineStageProtocol {
+
+    public var next: PipelineStageProtocol?
+    private let log: PipelineLogger
+    private lazy var attachmentRegex = NSRegularExpression("attachment; ?filename=([\"\\w.]+)",
+                                                           options: .caseInsensitive)
+
+    public init(logger: PipelineLogger? = nil) {
+        if let logger = logger {
+            self.log = logger
+        } else if #available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *) {
+            self.log = PipelineOSLogAdapter()
+        } else {
+            self.log = PipelineNSLogger()
+        }
+    }
+
+    public func onRequest(_ request: inout PipelineRequest) {
+        let req = request.httpRequest
+        log.debug("Request URL: \(req.url)")
+        log.debug("Request method: \(req.httpMethod.rawValue)")
+        log.debug("Request headers:")
+        for (header, value) in req.headers {
+            log.debug("    \(header): \(value)")
+        }
+        log.debug("Request body:")
+        log.debug(loggableBodyFor(headers: req.headers, bodyText: req.text))
+    }
+
+    public func onResponse(_ response: inout PipelineResponse) {
+        guard let res = response.httpResponse else {
+            log.debug("Failed to log response")
+            return
+        }
+
+        if let statusCode = res.statusCode {
+            log.debug("Response status: \(statusCode)")
+        }
+        log.debug("Response headers:")
+        for (header, value) in res.headers {
+            log.debug("    \(header): \(value)")
+        }
+        log.debug("Response content:")
+        log.debug(loggableBodyFor(headers: res.headers, bodyText: res.text))
+    }
+
+    public func onError(_ error: Error) -> Bool {
+        log.error(error.localizedDescription)
+        return false
+    }
+
+    private func loggableBodyFor(headers: HttpHeaders, bodyText: (String.Encoding) -> String?) -> String {
+        if let fileName = filenameFrom(header: headers[.contentDisposition]) {
+            return "File attachments: \(fileName)"
+        }
+
+        if let contentType = headers[.contentType] {
+            if contentType.lowercased().hasSuffix("octet-stream") {
+                return "Body contains binary data."
+            } else if contentType.lowercased().hasPrefix("image") {
+                return "Body contains image data."
+            }
+        }
+        return bodyText(.utf8) ?? "(empty)"
+    }
+
+    private func filenameFrom(header: String?) -> Substring? {
+        guard let header = header else { return nil }
+        guard let match = attachmentRegex.firstMatch(in: header) else { return nil }
+        let captures = match.capturedValues(from: header)
+        if captures.count > 1 {
+            return captures[1]
+        } else {
+            return nil
+        }
+    }
+}
+
+public class CurlFormattedRequestLoggingPolicy: PipelineStageProtocol {
```

Comment:
Is there a use case for this or was it a carry-over from Python?

Reply:
It's brand new for this implementation. In my experience debugging mobile applications it's incredibly useful to be able to have this output during development, and I always seem to end up writing / porting one of these to every mobile application I've worked on so I figured it would be simple to provide one. It would generally be enabled by providing this policy in addition to, rather than instead of, the regular logging policy.

Reply:
Cool, interested in checking it out